app/ante/cosmos/sigverify.go

```diff
@@ -0,0 +1,584 @@
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256r1"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/cosmos/cosmos-sdk/crypto/types/multisig"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/cosmos/cosmos-sdk/x/auth/migrations/legacytx"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+)
+
+var (
+	// simulation signature values used to estimate gas consumption
+	key                                            = make([]byte, secp256k1.PubKeySize)
+	simSecp256k1Pubkey                             = &secp256k1.PubKey{Key: key}
+	_                  authsigning.SigVerifiableTx = (*legacytx.StdTx)(nil) // assert StdTx implements SigVerifiableTx
+)
+
+func init() {
+	// This decodes a valid hex string into a sepc256k1Pubkey for use in transaction simulation
+	bz, _ := hex.DecodeString("035AD6810A47F073553FF30D2FCC7E0D3B1C0B74B61A1AAA2582344037151E143A")
+	copy(key, bz)
+	simSecp256k1Pubkey.Key = key
+}
+
+// SignatureVerificationGasConsumer is the type of function that is used to both
+// consume gas when verifying signatures and also to accept or reject different types of pubkeys
+// This is where apps can define their own PubKey
+type SignatureVerificationGasConsumer = func(meter sdk.GasMeter, sig signing.SignatureV2, params types.Params) error
+
+// SetPubKeyDecorator sets PubKeys in context for any signer which does not already have pubkey set
+// PubKeys must be set in context for all signers before any other sigverify decorators run
+// CONTRACT: Tx must implement SigVerifiableTx interface
+type SetPubKeyDecorator struct {
+	ak authante.AccountKeeper
+}
+
+func NewSetPubKeyDecorator(ak authante.AccountKeeper) SetPubKeyDecorator {
+	return SetPubKeyDecorator{
+		ak: ak,
+	}
+}
+
+func (spkd SetPubKeyDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	// skip publickkey set for oracle create-price message
+	if isOracleCreatePriceTx(tx) {
+		sigTx, ok := tx.(authsigning.SigVerifiableTx)
+		if !ok {
+			return ctx, sdkerrors.ErrTxDecode.Wrap("invalid tx type")
+		}
+		signers := sigTx.GetSigners()
+		pubKeys, err := sigTx.GetPubKeys()
+		if err != nil {
+			return ctx, err
+		}
+		for i, pk := range pubKeys {
+			// addrFromPubk, err := sdk.AccAddressFromBech32(sdk.AccAddress(pk).String())
+			if !bytes.Equal(signers[i], pk.Address()) {
+				return ctx, sdkerrors.ErrInvalidPubKey.Wrapf("pubKey does not match signer address %s with signer index: %d", signers[i], i)
+			}
+		}
+		return next(ctx, tx, simulate)
+	}
+
+	sigTx, ok := tx.(authsigning.SigVerifiableTx)
+	if !ok {
+		return ctx, sdkerrors.ErrTxDecode.Wrap("invalid tx type")
```

Comment:
the error is very simple, could be "invalid tx type, expected SigVerifiableTx"

Reply:
updated